docs/adr/0014-auth-for-settings-redesign.md

```diff
@@ -0,0 +1,96 @@
+# Authentication in Settings Redesign
+
+- Status: accepted
+- Deciders: Les Orchard, Lauren Zugai, Ben Bangert, Ryan Kelly
+- Date: 2020-04-15
+
+Issue: https://jira.mozilla.com/browse/FXA-1559
+
+## Context and Problem Statement
+
+The Settings Redesign app needs to query & mutate the same protected user data as the existing settings app hosted on content-server. This will require some form of authentication & authorization to manage that data.
+
+## Decision Drivers
+
+- Smooth UX
+- Security
+- Development velocity
+- Ease of integration
+
+## Considered Options
+
+- Authenticate via OAuth2 to use auth-server APIs
+- Authenticate via OAuth2 to use DB directly
+- Reuse existing sessionToken on content-server with auth-server APIs
+- Implement own login sessions
+
+## Decision Outcome
+
+Chosen option: "Authenticate via OAuth2 to use auth-server APIs", because:
+
+- It requires minimal changes to existing auth-server implementation and infrastructure.
+- It relies on an authentication mechanism with relatively well-known security properties.
+
+### Positive Consequences
+
+- The OAuth2 access token mechanism is better suited to this purpose than a novel scheme to share session token credentials.
+- The new Settings Redesign app can use existing auth-server APIs with minimal modifications to accept scoped OAuth2 access tokens. This can constrain most of the novelty in the project to the redesigned & reimplemented settings UX.
+
+### Negative Consequences
+
+- We will need to modify auth strategies in auth-server APIs, taking care not to affect existing usage in the production settings app.
+- We don't have an entirely greenfield project, which could be a bit of a drag. But, rediscovering lessons learned in existing code can also be a drag.
+
+## Pros and Cons of the Options
+
+### Authenticate via OAuth2 to use auth-server APIs
+
+We would like to someday deprecate the session token mechanism altogether. This could be a start.
+
+- Pros:
+  - Security properties of OAuth2 authn / authz are known and not entirely novel
+  - Auth-server APIs can be reused with small changes
+- Neutral:
+  - GraphQL server needs to proxy auth-server APIs
+- Cons:
+  - APIs on auth-server need auth strategy changes to accept OAuth tokens
+  - May initially require a hack similar to payments-server where a pre-generated access token is passed along
+
+### Authenticate via OAuth2 to use DB directly
+
+We would like to someday deprecate the session token mechanism altogether. We would also like to modernize around the sorts of things auth-server does. This could be a start.
+
+- Pros:
+  - Security properties of OAuth2 authn / authz are known and not entirely novel
+  - Can re-use at least the oauth APIs on auth-server to validate token stored in client
+- Neutral:
+  - GraphQL server can be implemented from scratch directly against DB
```

Comment:
Why can't the GQL server talk to the db despite what authentication is used? I'm not clear on why talking to the database is dependent on the authentication scheme.

Reply:
The GQL server *could* talk to the DB. Point is to *not* do that at first in order to avoid the scope of reimplementing auth-server logic at first.

Reply:
GraphQL Queries could mostly be implemented directly against the DB with very little overhead, I agree entirely that mutations involve quite a bit more logic and should all go to the auth-server initially.

Reply:
I think even some of the GET-only API have some logic in them that would be nice not to re-implement right now. And GQL queries that go straight to the DB will themselves need to verify the incoming OAuth access token. So, we'll need the access token for both direct DB accesses and auth-server API calls.

Reply:
I think they will need an OAuth token. My comment was mainly that as long as we're using an OAuth token, we could still go straight to the DB if desired. The ADR comment makes it sound like one of the OAuth token solutions doesn't allow for that.

Reply:
> The ADR comment makes it sound like one of the OAuth token solutions doesn't allow for that.

Which ADR comment? Should be the opposite: i.e. looking for a solution that allows re-use of existing auth-server APIs so we don't have to re-write everything. If we have direct DB access, we can pretty much do whatever

Reply:
Oh, the one this long thread is referring to as a neutral point. This neutral point is missing from the first choice above as well, but it could also hit the db directly if desired.